src/herder/TransactionQueue.cpp

```diff
@@ -123,91 +58,85 @@ TransactionQueue::tryAdd(TransactionFramePtr tx)
         return TransactionQueue::AddResult::ADD_STATUS_ERROR;
     }
 
-    auto map = findOrAdd(mPendingTransactions[0], tx->getSourceID());
-    map->addTx(tx);
-
+    auto& pendingForAccount = mPendingTransactions[tx->getSourceID()];
+    mSizeByAge[pendingForAccount.mAge]->inc();
+    pendingForAccount.mTotalFees += tx->getFeeBid();
+    pendingForAccount.mTransactions.emplace_back(tx);
     return TransactionQueue::AddResult::ADD_STATUS_PENDING;
 }
 
 void
 TransactionQueue::remove(std::vector<TransactionFramePtr> const& dropTxs)
 {
-    for (auto& m : mPendingTransactions)
+    for (auto const& tx : dropTxs)
     {
-        if (m.empty())
+        auto it = find(tx);
+        if (it.first == std::end(mPendingTransactions))
         {
             continue;
         }
 
-        std::unordered_set<std::shared_ptr<TxMap>> toRecalculate;
-
-        for (auto const& tx : dropTxs)
+        auto& txs = it.first->second.mTransactions;
+        auto feeBid =
+            std::accumulate(it.second, std::end(txs), 0,
+                            [](int fee, TransactionFramePtr const& tx) {
+                                return fee + tx->getFeeBid();
+                            });
+        it.first->second.mTotalFees -= feeBid;
+        txs.erase(it.second, std::end(txs));
```

Comment:
should those be banned as well?

Reply:
This function is used in two cases. First, when transaction in TransactionQueue became invalid for some reason. Second, when transaction has been applied on ledger.

I think in first case we can ban all removed transactions as well, but in second case it will result in `ADD_STATUS_TRY_AGAIN_LATER` instead of `ADD_STATUS_ERROR` when someone try to submit just applied transaction again. I don't think that it is a huge problem, or a problem at all.

So I think we should ban those as well in both cases. If you agree, I'll implement that shortly.

Reply:
yeah both cases are potentially bad for the validator, so banning seems to be the right strategy